File: src/lib/types/Album.ts
import type {ExternalUrls} from "./ExternalUrls";
import type {Image} from "./Image";
import type {Restrictions} from "./Restrictions";

export interface Album {
  album_type?: "album" | "single" | "compilation";
  total_tracks: number
  available_markets?: string[] | undefined
  external_urls: ExternalUrls
  href: string
  id: string
  images?: Image[]
  label?: string | undefined
  name: string
  release_date: string
  release_date_precision: "year" | "month" | "day"
  restrictions?: Restrictions | undefined
  type: "album"
  artists: {id: string, name: string}[]
  tracks: { id: string, name: string }[]
  genres?: string[] | undefined
  popularity?: number | undefined
  uri: string
  complete: boolean
}

export interface LibraryAlbum{
  album: {
    id: string,
    name: string,
    artists: {id: string, name: string}[]
    total_tracks: number
    images?: Image[]
  },
  added_at: string
  // This field is used when two libraries are compared, to indicate on which library the album is present.
  // Null indicates, it is present on both libraries
  library?: string | null
}

export function toAlbum(album: SpotifyApi.AlbumObjectFull): Album {
  return {
    album_type: album.album_type,
    total_tracks: album.total_tracks,
    available_markets: album.available_markets,
    external_urls: album.external_urls,
    href: album.href,
    id: album.id,
    images: album.images,
    label: album.label,
    name: album.name,
    release_date: album.release_date,
    release_date_precision: album.release_date_precision,
    restrictions: album.restrictions,
    type: album.type,
    artists: album.artists.map(a => ({id: a.id, name: a.name})),
    tracks: [],
    genres: album.genres,
    popularity: album.popularity,
    uri: album.uri,
    complete: false
  }
}

export function toLibraryAlbum(album: SpotifyApi.SavedAlbumObject): LibraryAlbum {
  return {
    album: {
      id: album.album.id,
      name: album.album.name,
      artists: album.album.artists.map(a => ({id: a.id, name: a.name})),
      total_tracks: album.album.total_tracks,
      images: album.album.images
    },
    added_at: album.added_at,
    library: null
  }
}

File: src/main/java/de/wittenbude/exportify/domain/entities/Album.java
package de.wittenbude.exportify.domain.entities;

import com.neovisionaries.i18n.CountryCode;
import de.wittenbude.exportify.domain.valueobjects.Copyright;
import de.wittenbude.exportify.domain.valueobjects.ExternalIDs;
import de.wittenbude.exportify.domain.valueobjects.Image;
import de.wittenbude.exportify.domain.valueobjects.SpotifyObjectType;
import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.Setter;
import lombok.experimental.Accessors;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.JdbcTypeCode;
import org.hibernate.proxy.HibernateProxy;
import org.hibernate.type.SqlTypes;

import java.time.Instant;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.UUID;

@Getter
@Setter
@Accessors(chain = true)
@Entity
public class Album {

    @Id
    @GeneratedValue
    private UUID id;

    @CreationTimestamp
    private Instant versionTimestamp;

    private String albumType;
    private Integer totalTracks;
    private String href;
    private String spotifyID;
    private String name;
    private String releaseDate;
    private ReleaseDatePrecision releaseDatePrecision;
    private String restrictions;
    private String uri;
    private ExternalIDs externalIDs;
    private String label;
    private Integer popularity;

    @Enumerated(EnumType.STRING)
    private SpotifyObjectType spotifyObjectType;

    @JdbcTypeCode(SqlTypes.JSON)
    private List<Copyright> copyrights;

    @JdbcTypeCode(SqlTypes.JSON)
    private List<Image> images;

    @JdbcTypeCode(SqlTypes.ARRAY)
    private List<CountryCode> availableMarkets;

    @JdbcTypeCode(SqlTypes.JSON)
    private Map<String, String> externalUrls;

    @JdbcTypeCode(SqlTypes.ARRAY)
    private List<String> genres;

    @JdbcTypeCode(SqlTypes.ARRAY)
    private List<String> spotifyArtistIDs;

    @JdbcTypeCode(SqlTypes.ARRAY)
    private List<String> spotifyTrackIDs;

    @Override
    public final boolean equals(Object o) {
        if (this == o) return true;
        if (o == null) return false;
        Class<?> oEffectiveClass = o instanceof HibernateProxy ? ((HibernateProxy) o).getHibernateLazyInitializer().getPersistentClass() : o.getClass();
        Class<?> thisEffectiveClass = this instanceof HibernateProxy ? ((HibernateProxy) this).getHibernateLazyInitializer().getPersistentClass() : this.getClass();
        if (thisEffectiveClass != oEffectiveClass) return false;
        Album album = (Album) o;
        return getId() != null && Objects.equals(getId(), album.getId());
    }

    @Override
    public final int hashCode() {
        return this instanceof HibernateProxy ? ((HibernateProxy) this).getHibernateLazyInitializer().getPersistentClass().hashCode() : getClass().hashCode();
    }

    @Getter
    @AllArgsConstructor
    public enum ReleaseDatePrecision {
        DAY,
        MONTH,
        YEAR
    }


}

